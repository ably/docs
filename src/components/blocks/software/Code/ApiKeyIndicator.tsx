import React from 'react';

const tooltip =
  'This code example uses a temporary key that is rate limited and expires in 4 hours. Sign in to Ably to use your API keys instead.';

const APIKeyIndicator = () => (
  <div className="docs-api-key-label" title={tooltip}>
    Api Key: Demo Only
  </div>
);

export default APIKeyIndicator;
