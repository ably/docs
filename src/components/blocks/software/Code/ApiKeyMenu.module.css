.container {
  width: 230px;
}
