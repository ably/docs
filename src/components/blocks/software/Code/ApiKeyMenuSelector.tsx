import React, { useEffect } from 'react';
import { DEFAULT_API_KEY_MESSAGE } from '.';
import APIKeyIndicator from './ApiKeyIndicator';
import APIKeyMenu, { APIKeyMenuProps, Option } from './ApiKeyMenu';

type APIKeyMenuSelectorProps = {
  dataContainsKey: boolean;
  signedIn: boolean;
  activeApiKey: Option;
} & APIKeyMenuProps;

const APIKeyMenuSelector = ({
  dataContainsKey,
  userApiKeys,
  setActiveApiKey,
  activeApiKey,
  signedIn = false,
}: APIKeyMenuSelectorProps) => {
  useEffect(() => {
    if (activeApiKey.value === DEFAULT_API_KEY_MESSAGE && userApiKeys) {
      const { name: label, whole_key: value } = userApiKeys[0].apiKeys[0];
      setActiveApiKey({
        label,
        value,
      });
    }
  }, [userApiKeys, activeApiKey, setActiveApiKey]);
  return dataContainsKey ? (
    signedIn && userApiKeys ? (
      <APIKeyMenu userApiKeys={userApiKeys} setActiveApiKey={setActiveApiKey} />
    ) : (
      <APIKeyIndicator />
    )
  ) : null;
};

export default APIKeyMenuSelector;
