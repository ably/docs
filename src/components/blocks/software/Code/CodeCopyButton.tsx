import React, { FunctionComponent as FC } from 'react';
import Icon from '@ably/ui/core/Icon';
import { ButtonWithTooltip } from 'src/components';
import { safeWindow } from 'src/utilities';
import { copyCodeBlockContentTracker } from 'src/external-scripts/google-tag-manager/events';
import { DEFAULT_LANGUAGE } from '../../../../../data/createPages/constants';

type Props = {
  content: string;
  language?: string;
};

const CodeCopyButton: FC<Props> = ({ content, language }) => {
  const handleCopy = () => {
    navigator.clipboard.writeText(content);
    const page = safeWindow.location.pathname;
    const contentIdentifier = content.slice(0, 10);
    copyCodeBlockContentTracker({
      copyCodeBlockContent: `docs - ${language ?? DEFAULT_LANGUAGE} ${page} ${contentIdentifier}`,
    });
  };

  return (
    <div className="absolute top-64 right-16">
      <ButtonWithTooltip tooltip="Copy" notification="Copied!" onClick={handleCopy} className="text-white">
        <Icon name="icon-gui-copy" size="1rem" color="white" />
      </ButtonWithTooltip>
    </div>
  );
};

export default CodeCopyButton;
