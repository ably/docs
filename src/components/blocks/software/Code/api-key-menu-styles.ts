import { SetStateAction } from 'react';
import { CSSObjectWithLabel, StylesConfig } from 'react-select';
import { gui, primary } from '../../../../styles/colors';

const SELECT_HEIGHT = '24px';

const optionStyles = (provided: CSSObjectWithLabel) => ({
  ...provided,
  width: '200px',
  border: 0,
  margin: 0,
  color: '#D9D9DA',
  backgroundColor: primary.charcoalGrey,
  boxShadow: 'none',
  fontFamily: `NEXT Book,Arial,Helvetica,sans-serif`,
  fontSize: '14px',
  '&:active': {
    ...provided[':active'],
    backgroundColor: gui.active,
  },
});

const controlStyles = (provided: CSSObjectWithLabel) => ({
  ...provided,
  width: '200px',
  border: 0,
  margin: 0,
  color: '#D9D9DA',
  backgroundColor: primary.charcoalGrey,
  boxShadow: 'none',
  fontFamily: `NEXT Book,Arial,Helvetica,sans-serif`,
  fontSize: '14px',
  lineHeight: 'var(--lh-snug)',
  height: SELECT_HEIGHT,
  minHeight: SELECT_HEIGHT,
});

const heightStyles = (provided: CSSObjectWithLabel) => ({
  ...provided,
  height: SELECT_HEIGHT,
});

export const selectMenuStyles: StylesConfig<SetStateAction<{ label: string; value: string }>, false> = {
  container: (provided: CSSObjectWithLabel) => ({
    ...provided,
    position: 'absolute',
    top: '0.25em',
    left: '0.25em',
    backgroundColor: primary.charcoalGrey,
    minHeight: SELECT_HEIGHT,
    height: SELECT_HEIGHT,
  }),
  valueContainer: heightStyles,
  indicatorSeparator: () => ({
    display: 'none',
  }),
  control: controlStyles,
  option: optionStyles,
  singleValue: optionStyles,
  groupHeading: optionStyles,
  menu: (provided: CSSObjectWithLabel) => ({
    ...provided,
    backgroundColor: primary.charcoalGrey,
    top: '1em',
  }),
  input: (provided: CSSObjectWithLabel) => ({
    ...provided,
    color: '#D9D9DA',
  }),
  dropdownIndicator: (provided: CSSObjectWithLabel) => ({
    ...provided,
    height: SELECT_HEIGHT,
    padding: '2px 8px 8px 8px',
    verticalAlign: 'center',
    color: '#D9D9DA',
    '&:hover': {
      color: 'white',
    },
  }),
};
