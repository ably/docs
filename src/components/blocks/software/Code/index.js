import React, { useMemo, useState } from 'react';
import PropTypes from 'prop-types';
import Html from '../../Html';

import { Light as SyntaxHighlighter } from 'react-syntax-highlighter';
import '@ably/ui/core/Code/component.css';
import '@ably/ui/core/styles.css';
import '../styles.css';

// Supported languages need to be imported here
// https://github.com/highlightjs/highlight.js/blob/master/SUPPORTED_LANGUAGES.md
import js from 'react-syntax-highlighter/dist/cjs/languages/hljs/javascript';
import ruby from 'react-syntax-highlighter/dist/cjs/languages/hljs/ruby';
import java from 'react-syntax-highlighter/dist/cjs/languages/hljs/java';
import python from 'react-syntax-highlighter/dist/cjs/languages/hljs/python';
import php from 'react-syntax-highlighter/dist/cjs/languages/hljs/php';
import bash from 'react-syntax-highlighter/dist/cjs/languages/hljs/bash';
import cs from 'react-syntax-highlighter/dist/cjs/languages/hljs/csharp';
import go from 'react-syntax-highlighter/dist/cjs/languages/hljs/go';
import xml from 'react-syntax-highlighter/dist/cjs/languages/hljs/xml';
import cpp from 'react-syntax-highlighter/dist/cjs/languages/hljs/cpp';
import dart from 'react-syntax-highlighter/dist/cjs/languages/hljs/dart';
import swift from 'react-syntax-highlighter/dist/cjs/languages/hljs/swift';
import objectivec from 'react-syntax-highlighter/dist/cjs/languages/hljs/objectivec';
import json from 'react-syntax-highlighter/dist/cjs/languages/hljs/json';
// Android Studio has an error when registering the language.

import languageLabels, { languageSyntaxHighlighterNames } from '../../../../maps/language';
import HtmlDataTypes from '../../../../../data/types/html';
import UserContext, { devApiKeysPresent } from '../../../../contexts/user-context';
import APIKeyMenuSelector from './ApiKeyMenuSelector';
import InlineCodeElement from './InlineCodeElement';
import CodeCopyButton from './CodeCopyButton';

export const DEFAULT_API_KEY_MESSAGE = '<loading API key, please wait>';

const SelectedLanguage = ({ language }) =>
  language ? <div className="docs-language-label">{language.label}</div> : null;

SelectedLanguage.propTypes = {
  language: PropTypes.oneOfType([PropTypes.string, PropTypes.object]),
};

SyntaxHighlighter.registerLanguage(languageSyntaxHighlighterNames.javascript.key, js);
SyntaxHighlighter.registerLanguage(languageSyntaxHighlighterNames.java.key, java);
SyntaxHighlighter.registerLanguage(languageSyntaxHighlighterNames.ruby.key, ruby);
SyntaxHighlighter.registerLanguage(languageSyntaxHighlighterNames.python.key, python);
SyntaxHighlighter.registerLanguage(languageSyntaxHighlighterNames.php.key, php);
SyntaxHighlighter.registerLanguage(languageSyntaxHighlighterNames.shell.key, bash);
SyntaxHighlighter.registerLanguage(languageSyntaxHighlighterNames.csharp.key, cs);
SyntaxHighlighter.registerLanguage(languageSyntaxHighlighterNames.go.key, go);
SyntaxHighlighter.registerLanguage(languageSyntaxHighlighterNames.html.key, xml);
SyntaxHighlighter.registerLanguage(languageSyntaxHighlighterNames.cpp.key, cpp);
SyntaxHighlighter.registerLanguage(languageSyntaxHighlighterNames.dart.key, dart);
SyntaxHighlighter.registerLanguage(languageSyntaxHighlighterNames.swift.key, swift);
SyntaxHighlighter.registerLanguage(languageSyntaxHighlighterNames.objc.key, objectivec);
SyntaxHighlighter.registerLanguage(languageSyntaxHighlighterNames.json.key, json);

const multilineRegex = /\r|\n/gm;

const Code = ({ data, attribs }) => {
  const [activeApiKey, setActiveApiKey] = useState({
    label: DEFAULT_API_KEY_MESSAGE,
    value: DEFAULT_API_KEY_MESSAGE,
  });
  const isString = data.length === 1 && data[0].type === HtmlDataTypes.text;
  const hasRenderableLanguages = isString && attribs && attribs.lang;
  const hasMultilineText = isString && multilineRegex.test(data[0].data);

  const content = data[0]?.data ?? '';
  /**
   * Refer to Decision Record:
   * https://ably.atlassian.net/wiki/spaces/ENG/pages/2070053031/DR9+API+Keys+vs+tokens+vs+authUrls+in+docs+code+snippets#Recommendation
   * Referenced on ticket:
   * https://ably.atlassian.net/browse/EDX-49
   */
  const contentWithObfuscatedKey = useMemo(
    () =>
      content.replace && content.replace(/{{API_KEY}}/g, '*********************************************************'),
    [content, activeApiKey],
  );
  const contentWithKey = useMemo(
    () => content.replace && content.replace(/{{API_KEY}}/g, activeApiKey.value),
    [content, activeApiKey],
  );

  if (hasRenderableLanguages || hasMultilineText) {
    const dataContainsKey = attribs['data-contains-key'] === 'true';
    const displayLanguage =
      attribs.lang && languageSyntaxHighlighterNames[attribs.lang]
        ? languageSyntaxHighlighterNames[attribs.lang]
        : languageSyntaxHighlighterNames['plaintext'];
    return (
      <div {...attribs} className="p-32 overflow-auto relative" language={languageLabels[attribs.lang]}>
        <UserContext.Consumer>
          {(value) => (
            <APIKeyMenuSelector
              dataContainsKey={dataContainsKey}
              userApiKeys={process.env.GATSBY_DOCS_API_KEYS ? devApiKeysPresent : value.apiKeys.data}
              setActiveApiKey={setActiveApiKey}
              activeApiKey={activeApiKey}
              signedIn={!!value.sessionState.signedIn || !!process.env.GATSBY_DOCS_SIGNED_IN}
            />
          )}
        </UserContext.Consumer>
        <SelectedLanguage language={displayLanguage} />
        <UserContext.Consumer>
          {(value) => (
            <SyntaxHighlighter
              className="ui-text-code"
              style={{ hljs: { background: 'inherit', fontSize: `var(--fs-code)`, lineHeight: `var(--lh-loose)` } }}
              language={displayLanguage.key}
            >
              {dataContainsKey
                ? value.sessionState.signedIn || !!process.env.GATSBY_DOCS_SIGNED_IN
                  ? contentWithObfuscatedKey
                  : contentWithKey
                : content}
            </SyntaxHighlighter>
          )}
        </UserContext.Consumer>
        <CodeCopyButton content={contentWithKey} />
      </div>
    );
  }
  return (
    <InlineCodeElement {...attribs}>
      <Html data={data} />
    </InlineCodeElement>
  );
};

Code.propTypes = {
  data: PropTypes.array,
  attribs: PropTypes.object,
};

export default Code;
