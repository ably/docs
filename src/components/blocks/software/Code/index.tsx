import React, { useMemo, useState } from 'react';
import { some } from 'lodash/fp';
// @ts-ignore
import languagesRegistry from '@ably/ui/core/utils/syntax-highlighter-registry';

import Html from 'src/components/blocks/Html';
import UserContext from 'src/contexts/user-context';

import HtmlDataTypes from '../../../../../data/types/html';
import { API_KEY_DATA_ATTRIBUTE } from '../../../../../data/html-parser/add-info-to-codeblocks/codeblock-api-key-info';
import { RANDOM_CHANNEL_NAME_DATA_ATTRIBUTE } from '../../../../../data/html-parser/add-info-to-codeblocks/codeblock-random-channel-name';

import APIKeyMenuSelector from './ApiKeyMenuSelector';
import InlineCodeElement from './InlineCodeElement';
import CodeCopyButton from './CodeCopyButton';
import { MultilineCodeContent } from './MultilineCodeContent';
import { getRandomChannelName } from './get-random-channel-name';

import { NestedHtmlComponentProps } from 'src/components/html-component-props';
import { extractCodeStringsFromContent } from './extract-code-strings-from-content';
import { BASH_LANGUAGE } from '../../../../../data/createPages/constants';

const API_KEY_LENGTH = 5;
export const DEFAULT_API_KEY_MESSAGE = '<loading API key, please wait>';

export const multilineRegex = /\r|\n/gm;

const alternativeLanguageRegistry: Record<string, string> = {
  objc: 'objectivec',
  nodejs: 'javascript',
  react: 'jsx',
  dotnet: 'cs',
  csharp: 'cs',
  flutter: 'dart',
};

const retrieveFromAlternativeLanguageRegistry = (key?: string) =>
  key ? alternativeLanguageRegistry[key] : BASH_LANGUAGE;

const Code = ({ data, attribs }: NestedHtmlComponentProps<'div'>) => {
  const [activeApiKey, setActiveApiKey] = useState({
    label: DEFAULT_API_KEY_MESSAGE,
    value: DEFAULT_API_KEY_MESSAGE,
  });

  const isString = some((child) => child.type === HtmlDataTypes.text, data);
  const hasRenderableLanguages = isString && attribs && attribs.lang;

  const dataContainsKey = attribs?.[`data-contains-${API_KEY_DATA_ATTRIBUTE}`] === 'true';
  const dataContainsRandomChannelName = attribs?.[`data-contains-${RANDOM_CHANNEL_NAME_DATA_ATTRIBUTE}`] === 'true';

  const content = extractCodeStringsFromContent(data)
    .join('')
    .replace(/<\/?code>/g, '@'); // Don't nest <code> inside <code> components; it's probably textile-js mis-interpreting Objective-C code.
  const contentWithRandomChannelName = useMemo(
    () =>
      dataContainsRandomChannelName ? content.replace(/{{RANDOM_CHANNEL_NAME}}/g, getRandomChannelName()) : content,
    [content, dataContainsRandomChannelName],
  );

  const hasMultilineText = isString && multilineRegex.test(content);
  /**
   * Refer to Decision Record:
   * https://ably.atlassian.net/wiki/spaces/ENG/pages/2070053031/DR9+API+Keys+vs+tokens+vs+authUrls+in+docs+code+snippets#Recommendation
   * Referenced on ticket:
   * https://ably.atlassian.net/browse/EDX-49
   */
  const contentWithObfuscatedKey = useMemo(() => {
    const firstSectionApiKey = activeApiKey.value.substring(0, activeApiKey.value.indexOf(':') + 1);
    const displayApiKey = firstSectionApiKey ? firstSectionApiKey : activeApiKey.value.substring(0, 10) + ':';
    return dataContainsKey
      ? contentWithRandomChannelName.replace(
          /{{API_KEY}}/g,
          `${displayApiKey}${new Array(API_KEY_LENGTH + 1).join('*')}`,
        )
      : contentWithRandomChannelName;
  }, [contentWithRandomChannelName, activeApiKey, dataContainsKey]);
  const contentWithKey = useMemo(
    () =>
      dataContainsKey
        ? contentWithRandomChannelName.replace(/{{API_KEY}}/g, activeApiKey.value)
        : contentWithRandomChannelName,
    [contentWithRandomChannelName, activeApiKey, dataContainsKey],
  );

  if (hasRenderableLanguages || hasMultilineText) {
    const languageInRegistry = (languagesRegistry as { key: string; label: string; module: unknown }[]).find(
      (languageData) => languageData.key === attribs?.lang,
    );
    const displayLanguage =
      attribs?.lang && languageInRegistry
        ? languageInRegistry.key
        : retrieveFromAlternativeLanguageRegistry(attribs?.lang);

    return (
      <>
        <div {...attribs} className="overflow-auto relative p-16" data-testid="code-block">
          <UserContext.Consumer>
            {(value) => (
              <MultilineCodeContent
                signedIn={value.sessionState.signedIn}
                dataContainsKey={dataContainsKey}
                content={contentWithRandomChannelName}
                contentWithKey={contentWithKey}
                contentWithObfuscatedKey={contentWithObfuscatedKey}
                language={displayLanguage}
              />
            )}
          </UserContext.Consumer>
        </div>
        <UserContext.Consumer>
          {(value) => (
            <APIKeyMenuSelector
              dataContainsKey={dataContainsKey}
              apps={value.apps}
              setActiveApiKey={setActiveApiKey}
              activeApiKey={activeApiKey}
              signedIn={!!value.sessionState.signedIn || !!process.env.GATSBY_DOCS_SIGNED_IN}
            />
          )}
        </UserContext.Consumer>
        <CodeCopyButton
          language={attribs?.lang}
          content={dataContainsKey ? contentWithKey : contentWithRandomChannelName}
        />
      </>
    );
  }
  return (
    <InlineCodeElement {...attribs}>
      <Html data={data} />
    </InlineCodeElement>
  );
};

export default Code;
