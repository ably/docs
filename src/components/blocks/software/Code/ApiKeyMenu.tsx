import React, { Dispatch, useState } from 'react';
import Select from 'react-select';
import { DEFAULT_API_KEY_MESSAGE } from '.';
import { SmallMenuLabel } from '../../../Menu/Label';
import { selectMenuStyles } from './api-key-menu-styles';

const VALUE_PREFIX = 'API Key: ';

export type Option = {
  label: string;
  value: string;
};

type AppApiKey = {
  name: string;
  whole_key: string;
};

export type UserApiKey = {
  name: string;
  url: string;
  apiKeys: AppApiKey[];
};

const errorOption = {
  label: 'No API keys found',
  value: 'Placeholder: replace with default API Key',
};

const makeLabel = (apiKey: AppApiKey) => `${VALUE_PREFIX}${apiKey.name}`;

export type APIKeyMenuProps = {
  userApiKeys: UserApiKey[];
  setActiveApiKey: Dispatch<React.SetStateAction<Option>>;
};

const APIKeyMenu = ({ userApiKeys, setActiveApiKey }: APIKeyMenuProps) => {
  const initialApiKeyOption = userApiKeys[0].apiKeys;

  const [value, setValue] = useState(
    initialApiKeyOption
      ? {
          label: `${makeLabel(initialApiKeyOption[0])}`,
          value: initialApiKeyOption[0].whole_key,
        }
      : errorOption,
  );

  const options = userApiKeys.map((userApiKey) => ({
    label: userApiKey.name,
    options: userApiKey.apiKeys.map((appApiKey) => ({
      label: makeLabel(appApiKey),
      value: appApiKey.whole_key,
    })),
  }));
  return (
    <>
      <SmallMenuLabel>API Key:</SmallMenuLabel>
      <Select
        value={value}
        onChange={(value) => {
          setValue(value ?? errorOption);
          setActiveApiKey(value ?? { label: DEFAULT_API_KEY_MESSAGE, value: DEFAULT_API_KEY_MESSAGE });
        }}
        options={options}
        styles={selectMenuStyles}
      />
    </>
  );
};

export default APIKeyMenu;
