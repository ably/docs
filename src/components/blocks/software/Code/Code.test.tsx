import React from 'react';
import { render, screen } from '@testing-library/react';
import Code from '.';

const rubyProps = {
  data: [
    {
      data: "channel = ably.channels.get('quickstart')\nchannel.publish 'greeting', 'hello!'",
      name: 'text',
      type: 'text',
    },
  ],
  attribs: { lang: 'ruby' },
};

const shProps = {
  data: [
    {
      data: `curl https://rest.ably.io/channels/{{RANDOM_CHANNEL_NAME}}/publish \\\n  --user "{{API_KEY}}" \\\n  --data "name=greeting&data=Hello"`,
      name: 'text',
      type: 'text',
    },
  ],
  attribs: { lang: 'sh' },
};

describe('<Code />', () => {
  it('Renders Code elements with language', () => {
    const { container } = render(<Code {...rubyProps} />);
    expect(container).toMatchSnapshot();
  });

  it('Renders Code elements without language', () => {
    const { container } = render(<Code {...shProps} attribs={{ lang: undefined }} />);
    expect(container.firstChild?.firstChild).toHaveClass('language-plaintext');
  });

  it('Updates the codeblock when language is switched', () => {
    const { rerender } = render(<Code {...rubyProps} />);

    expect(screen.getByTestId('code-block')).toMatchSnapshot();

    rerender(<Code {...shProps} />);
    expect(screen.getByTestId('code-block')).toMatchSnapshot();
  });
});
