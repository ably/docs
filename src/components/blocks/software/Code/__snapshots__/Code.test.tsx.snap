// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Code /> Renders Code elements with language 1`] = `
<div>
  <div
    class="overflow-auto relative p-16"
    data-testid="code-block"
    lang="ruby"
  >
    <code
      class="p-32 language-ruby"
      style="white-space: pre-wrap;"
    >
      channel 
      <span
        class="token operator"
      >
        =
      </span>
       ably
      <span
        class="token punctuation"
      >
        .
      </span>
      channels
      <span
        class="token punctuation"
      >
        .
      </span>
      get
      <span
        class="token punctuation"
      >
        (
      </span>
      <span
        class="token string-literal"
      >
        <span
          class="token string"
        >
          'quickstart'
        </span>
      </span>
      <span
        class="token punctuation"
      >
        )
      </span>
      
channel
      <span
        class="token punctuation"
      >
        .
      </span>
      publish 
      <span
        class="token string-literal"
      >
        <span
          class="token string"
        >
          'greeting'
        </span>
      </span>
      <span
        class="token punctuation"
      >
        ,
      </span>
       
      <span
        class="token string-literal"
      >
        <span
          class="token string"
        >
          'hello!'
        </span>
      </span>
    </code>
  </div>
  <div
    class="absolute top-14 right-16"
  >
    <div
      class="relative"
    >
      <button
        class="button text-white"
        type="button"
      >
        <svg
          class="white "
          style="width: 1rem; height: 1rem;"
        >
          <use
            xlink:href="#sprite-icon-gui-copy"
          />
        </svg>
      </button>
      <div
        class="tooltipClass"
        role="tooltip"
      >
        <span>
          Copy
        </span>
      </div>
      <div
        class="notificationClass"
      >
        <span>
          Copied!
        </span>
      </div>
    </div>
  </div>
</div>
`;

exports[`<Code /> Updates the codeblock when language is switched 1`] = `
<div
  class="overflow-auto relative p-16"
  data-testid="code-block"
  lang="ruby"
>
  <code
    class="p-32 language-ruby"
    style="white-space: pre-wrap;"
  >
    channel 
    <span
      class="token operator"
    >
      =
    </span>
     ably
    <span
      class="token punctuation"
    >
      .
    </span>
    channels
    <span
      class="token punctuation"
    >
      .
    </span>
    get
    <span
      class="token punctuation"
    >
      (
    </span>
    <span
      class="token string-literal"
    >
      <span
        class="token string"
      >
        'quickstart'
      </span>
    </span>
    <span
      class="token punctuation"
    >
      )
    </span>
    
channel
    <span
      class="token punctuation"
    >
      .
    </span>
    publish 
    <span
      class="token string-literal"
    >
      <span
        class="token string"
      >
        'greeting'
      </span>
    </span>
    <span
      class="token punctuation"
    >
      ,
    </span>
     
    <span
      class="token string-literal"
    >
      <span
        class="token string"
      >
        'hello!'
      </span>
    </span>
  </code>
</div>
`;

exports[`<Code /> Updates the codeblock when language is switched 2`] = `
<div
  class="overflow-auto relative p-16"
  data-testid="code-block"
  lang="sh"
>
  <code
    class="p-32 language-plaintext"
    style="white-space: pre-wrap;"
  >
    channel = ably.channels.get('quickstart')
channel.publish 'greeting', 'hello!'
  </code>
</div>
`;
